src/global.c

```diff
@@ -136,42 +103,58 @@ static void shutdown_common(void)
  */
 #if defined(GIT_THREADS) && defined(GIT_WIN32)
 
-static DWORD _fls_index;
-static volatile LONG _mutex = 0;
+static volatile LONG init_mutex = 0;
 
-static void WINAPI fls_free(void *st)
+GIT_INLINE(int) mutex_lock(void)
 {
-	git__global_state_cleanup(st);
-	git__free(st);
+	while (InterlockedCompareExchange(&init_mutex, 1, 0)) { Sleep(0); }
+	return 0;
 }
 
-static int synchronized_threads_init(void)
+GIT_INLINE(int) mutex_unlock(void)
 {
-	int error;
+	InterlockedExchange(&init_mutex, 0);
+	return 0;
+}
 
-	if ((_fls_index = FlsAlloc(fls_free)) == FLS_OUT_OF_INDEXES)
-		return -1;
+#elif defined(GIT_THREADS) && defined(_POSIX_THREADS)
 
-	error = init_common();
+static pthread_mutex_t init_mutex = PTHREAD_MUTEX_INITIALIZER;
 
-	return error;
+GIT_INLINE(int) mutex_lock(void)
```

Comment:
We now have essentially two different implementations of mutexes. Is the "thread-utils.h" implementation dependent on data we haven't initialized yet?

Reply:
This was poorly named and a leftover from prior refactorings.  On Windows, we use a spinlock here - we don't use `git_mutex` (which maps to a `CriticalSection`).  This is efficient and has a static initializer (since it operates on a `LONG`).

On Unix, however, we have to use a pthread mutex, since that gives us a static initializer.  (We can't call `pthread_something_init` since the goal of this function is to manage our initialization lifetime.)

I added comments and disambiguated the names to help show what's going on here and why.